### PR TITLE
Markdown

### PR DESCRIPTION
**Pull request by neuroscr**
Work in progress, do not merge. Posting per Maxim's request.

Public chat works
Note to self works
conversation last message works
outgoing message works in somecases

still need to redo how private conversation work with it; going to nothing with markdown on the senders side and move the detection to the receivers side

also needs some cleanup ofc

supports the following markdown:
```
- *em*
- _strong_ 
- `code` 
- ```multilinecode``` 
- 4 spaces then code
- > blockquotes
- - lists
```

does not handle nesting of em in strong and visa versa
plan to add more from https://support.discordapp.com/hc/en-us/articles/210298617-Markdown-Text-101-Chat-Formatting-Bold-Italic-Underline-

### CODE REVIEW

**Comment by neuroscr** [comment]
I do want to add some more markdown elements per https://support.discordapp.com/hc/en-us/articles/210298617-Markdown-Text-101-Chat-Formatting-Bold-Italic-Underline-

**Comment by neuroscr** [comment]
Looks like "grunt unit tests" are failing, any ideas on how to include `libloki_markdown.js` for those?

File: js/modules/loki_public_chat_api.js [hunk 1]
```diff
@@ -561,6 +580,17 @@ class LokiPublicChannelAPI {
           if (!from) {
             ({ from } = noteValue);
           }
+
+          // check for markdown
+          if (markdown) {
+            // DO NOT ALLOW users/servers to slip in their own custom HTML
+            const file = mdEngine.processSync(markdown);
+            text = String(file);
```
Comment (sachaaaaa) [inline js/modules/loki_public_chat_api.js]: Is there any way we can do the markdown processing at a lower level? There seems to be some duplicated logic between public and private messages.
Why not emit the raw message here and leave the markdown handling to the message.js model?

File: js/modules/loki_public_chat_api.js [hunk 2]
```diff
@@ -518,9 +525,12 @@ class LokiPublicChannelAPI {
   }
 
   async pollOnceForMessages() {
+    // mark sure send() can access this object instance
+    const ref = this;
```
Comment (sachaaaaa) [inline js/modules/loki_public_chat_api.js]: This looks like a binding issue?
Where is that function `send()` that needs to access this instance?

File: js/models/messages.js [hunk 3]
```diff
@@ -715,16 +726,29 @@
         return null;
       }
 
+      let processedText = text;
+      if (hasMarkdown(text)) {
+        // if we detect markDown, convert to HTML
+        processedText = window.markdoneHTMLConverter(text);
```
Comment (neuroscr) [inline js/models/messages.js]: argh I did here too!

File: js/modules/loki_public_chat_api.js (hunk 1 above)
Comment (neuroscr) [reply js/modules/loki_public_chat_api.js]: Mainly to keep it compatible with the appdotnet ecosystem (like the irc bridges, etc). This requires no markdown in the `text` field. The duplicate logic is contained to the libloki_markdown library itself.

File: js/modules/loki_public_chat_api.js (hunk 2 above)
Comment (neuroscr) [reply js/modules/loki_public_chat_api.js]: remanent from a previous version, I think we can remove these 2 lines

File: js/modules/loki_public_chat_api.js [hunk 4]
```diff
@@ -534,23 +544,32 @@ class LokiPublicChannelAPI {
...
         let timestamp = new Date(adnMessage.created_at).getTime();
         // pubKey lives in the username field
         let from = adnMessage.user.name;
         let quote = null;
+        let markdown = null;
         if (adnMessage.is_deleted) {
           return;
         }
+        let text;
+        ({ text } = adnMessage);
```
[the remark this answers is not in the document]
Comment (neuroscr) [reply js/modules/loki_public_chat_api.js]: I think I was given some grief on this pattern `{ field } =obj;` where I needed parenthesis, and couldn't get let to work with parenthesis. If this works, awesome, this is what I was looking for.

File: js/modules/loki_public_chat_api.js [hunk 5]
```diff
@@ -534,23 +544,32 @@ class LokiPublicChannelAPI {
...
-
     if (!res.err && res.response) {
       let receivedAt = new Date().getTime();
+      const mdEngine = remark()
+        // DO NOT ALLOW users/servers to inject their own custom HTML
+        // not sure I trust this
+        // this does allow `<b>bold example</b>` to work
+        .use(remarkStrip)
+        .use(remarkRecommended) // better cross browser support?
+        .use(remarkHtml); // and finally convert to HTML
```
[the remark this answers is not in the document]
Comment (neuroscr) [reply js/modules/loki_public_chat_api.js]: yes, I think this was done before I understood how preload.js/window worked

**Comment by Mikunj** [comment]
Going to close this for now to clean up the PRs.

**Pull request closed without merging**